### PR TITLE
[ThinLTO] Asynchronous caching

### PR DESCRIPTION
**Pull request by jansvoboda11**
This patch makes ThinLTO caching asynchronous, solving the problem of IO-bound cache operations blocking the thread pool for CPU-bound codegen.

We spawn asynchronous cache lookup queries for all modules at the start and then start proactively performing the codegen locally. Each module has both of these tasks racing. In the normal mode, we pick the winner to be written as the final object file. With deteminism checks on, we wait for both of these tasks to finish to compare their results.

Our current implementation of the remote caching service doesn't like concurrent read and write requests, so we wait with cache writes until all reads have finished.

We have a cancellation mechanism in place so that even with particularly slow remote cache, we're not much slower than with no cache at all.

Below are my wall-time measurements of the parallel part of `ThinLTOCodeGenerator::run()`, optimizing 3019 modules using our remote caching service:

| Cache  | Miss/Hit | Before | After |
| ------ | -------- | ------ | ----- |
| File   | Miss     |  74.3s | 75.4s |
| Remote | Miss     |  83.3s | 78.7s |
| Remote | Hit      |  31.1s | 22.4s |

Before this patch, using a particularly slow cache would result in linear unbounded growth of the wall-time, while the worst case after this patch is essentially an empty local cache + some timeout constant (currently hard-coded to 5s).

### CODE REVIEW

**Comment by jansvoboda11** [comment]
Just noting that there are still some parts that needs more polish. I'm looking for high-level comments and discussion on whether this level of additional complexity makes sense.

**Comment by jansvoboda11** [comment]
Both @cachemeifyoucan and @akyrtzi wanted to avoid diverging even more than we currently do from the upstream ThinLTO implementation. The newest commit makes the changes to `ThinLTOCodeGenerator.cpp` less spread out. This should also make it easier to remove the existing diff from upstream for the file-based caching and leave us mostly just additive changes to support the async use-case.

**Review by akyrtzi — APPROVED** [review]

File: llvm/lib/LTO/ThinLTOCodeGenerator.cpp [hunk 1]
```diff
@@ -1557,6 +1761,273 @@ void ThinLTOCodeGenerator::run() {
...
+            if (!CacheHit)
+              EnqueueCacheWrite(IndexCount);
+            else
+              EnqueueCacheWrite(-1);
+          }
+        }
+      });
+    }
+
+    // Parallel optimizer + codegen
```
Comment (cachemeifyoucan) [inline llvm/lib/LTO/ThinLTOCodeGenerator.cpp:1915]: Is this block same as blow?

File: llvm/lib/LTO/ThinLTOCodeGenerator.cpp [hunk 2]
```diff
@@ -1364,6 +1526,48 @@ ThinLTOCodeGenerator::writeGeneratedObject(StringRef OutputPath,
   return std::string(OutputPath);
 }
 
+template <class T>
```
Comment (cachemeifyoucan) [inline llvm/lib/LTO/ThinLTOCodeGenerator.cpp:1529]: Nit: Move to a local header in `lib/LTO`

**Review by cachemeifyoucan — COMMENTED** [review]
Some small nit picks. Diff is still quite large but seems more manageable.

File: llvm/lib/LTO/ThinLTOCodeGenerator.cpp (hunk 1 above)
Comment (jansvoboda11) [reply llvm/lib/LTO/ThinLTOCodeGenerator.cpp:1915]: No, this block is pretty different from the existing logic. Only the chunk of code between line 1960 and 1986 is identical, and I think it's still better to duplicate that in order to make the downstream diff more self-contained.

File: llvm/lib/LTO/ThinLTOCodeGenerator.cpp (hunk 2 above)
Comment (jansvoboda11) [reply llvm/lib/LTO/ThinLTOCodeGenerator.cpp:1529]: It's only used in this .cpp file, so I'd prefer to keep it here to reduce the spread of our downstream diff.

**Pull request merged**